### PR TITLE
Implementing stop feature on workflow

### PR DESCRIPTION
**Pull request by patkivikram**
## Description

This PR implements the stop functionality on the workflow. When stop is invoked on the workflow, dagger will invoke the stop method's implemented in the currently executing tasks

## Type of Change

- [ ] Bug Fix
- [x ] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [ ] Other (please describe)

## Checklist

- [ x] I have read the [contributing guidelines](https://github.com/wayfair-incubator/dagger/blob/main/CONTRIBUTING.md)
- [ x] Existing issues have been referenced (where applicable)
- [x ] I have verified this change is not present in other open pull requests
- [ x] Functionality is documented
- [x ] All code style checks pass
- [ x] New code contribution is covered by automated tests
- [x ] All new and existing tests pass

### CODE REVIEW

File: dagger/tasks/task.py [hunk 1]
```diff
@@ -945,7 +945,11 @@ class INonLeafNodeTask(ITask[KT, VT], abc.ABC):
     task_type: str = TaskType.SUB_DAG.name
 
     async def stop(self) -> None:
-        pass
+        self.status = TaskStatus(
+            code=TaskStatusEnum.STOPPED.name,
+            value=TaskStatusEnum.STOPPED.value,
+        )
+        await super().stop()
```
Comment (JeffreyGreenII) [inline dagger/tasks/task.py:952]: Is this necessary if the parent will set the status of the child tasks?

File: requirements-test.txt [hunk 2]
```diff
@@ -10,4 +10,4 @@ pytest-cov==4.0.0
 pytest-mock==3.10.0
 pytest-asyncio==0.14.0
 pytest-runner==4.2
-confluent_kafka==1.9.2
```
Comment (JeffreyGreenII) [inline requirements-test.txt:13]: Why are we switching libraries?

File: dagger/modeler/builder_helper.py [hunk 3]
```diff
@@ -141,7 +141,7 @@ def generic_command_task_builder(
             value_serializer=value_serializer,
         )
         command_task_builder = KafkaCommandTaskTemplateBuilder(self.app)
-        command_task_builder.set_topic(command_topic)
+        command_task_builder.set_topic(command_topic)  # type: ignore
```
Comment (JeffreyGreenII) [inline dagger/modeler/builder_helper.py:144]: Why do we need to ignore the type here now?

File: dagger/tasks/task.py [hunk 4]
```diff
@@ -1311,6 +1315,23 @@ class DefaultTemplateDAGInstance(ITemplateDAGInstance[str, str]):
...
 
+    async def stop(self) -> None:
+        remaining_tasks: Optional[List[ITask]] = await self.get_remaining_tasks(
+            next_dag_id=self.root_dag, workflow_instance=self, tasks=[]  # type: ignore
+        )  # type: ignore
+        if remaining_tasks:
+            for task in remaining_tasks:
+                if task.status.code == TaskStatusEnum.EXECUTING.name:
+                    await task.stop()
+                if task.status.code not in TERMINAL_STATUSES:
```
Comment (cmwindeknecht) [inline dagger/tasks/task.py:1326]: Just so I understand:
- Executing tasks need to have super.stop() called
- Not started / other non-terminal statuses do not need super.stop() / only need to have their status updated to STOPPED
- Terminal statuses are left as is

File: dagger/tasks/task.py (hunk 1 above)
Comment (cmwindeknecht) [reply dagger/tasks/task.py:952]: DWR: You could move the logic for checking status here and just call task.stop() on all remaining tasks in the parent dag - not sure why but that seems cleaner to me 😅 
```
    async def stop(self) -> None:
        if task.status.code not in TERMINAL_STATUSES:
          self.status = TaskStatus(
              code=TaskStatusEnum.STOPPED.name,
              value=TaskStatusEnum.STOPPED.value,
          )
        if task.status.code == TaskStatusEnum.EXECUTING.name:
           await super().stop()
```
And then just 
```
        if remaining_tasks:
            for task in remaining_tasks:
                task.stop()
```

File: dagger/modeler/builder_helper.py (hunk 3 above)
Comment (patkivikram) [reply dagger/modeler/builder_helper.py:144]: TopicT to Topic conversion

File: dagger/tasks/task.py (hunk 4 above)
Comment (patkivikram) [reply dagger/tasks/task.py:1326]: correct!

File: requirements-test.txt (hunk 2 above)
Comment (patkivikram) [reply requirements-test.txt:13]: this was not used at all

**Pull request merged**